### DISCUSSION TITLE
Is there a way to get the underlying sqlx pool?

### DISCUSSION THREAD

**Discussion by nick-kang** [SeaQL/sea-orm#1564] [Q&A] [answered]
There are some queries that are easier to write in raw sql and I would like the compile-time guarantees of sqlx. Is there a way to get the sqlx Pool out of `sea_orm::DatabaseConnection`?

My current workaround is just to instantiate two sqlx pools: one to use with sqlx queries and another that is used to instantiate `sea_orm::DatabaseConnection`.

**Comment by tyt2y3** [comment, marked as answer]
Yes, we added that recently behind the feature flag `sea-orm-internal`

https://github.com/SeaQL/sea-orm/blob/32fc8e236cd48cd935516516aa6d537ad742b78e/src/database/db_connection.rs#L412-L452

https://docs.rs/sea-orm/latest/sea_orm/entity/prelude/enum.DatabaseConnection.html#method.get_mysql_connection_pool

**Reply by nick-kang** [reply]
Awesome! Thank you!

**Reply by k-bx** [reply]
@tyt2y3 is there anything similar for `DatabaseTransaction` to get underlying transaction for sqlx?

**Reply by nick-kang** [reply]
@k-bx - Were you able to figure something out? We're trying to do the same thing. Would like to access the underlying sqlx transaction in `DatabaseTransaction` so we can run `sqlx::query_as!` and get the inline sql type checking.

Right now, we're just doing `transaction.execute(Statement::from_sql_and_values(...))`, which isn't ideal because there's no type checking.

**Reply by tyt2y3** [reply]
Currently there isn't, I'd suggest you to obtain the SQLx pool via the `internal` API then open a transaction using that connection.

However, other than that a similar API for getting the underlying transaction connection, e.g. `get_mysql_transaction(&mut self) -> Result<PoolConnection<sqlx::MySql>>` can be added (should be straightforward, PR is welcome), but it is only valid when `self.open == true`.

**Reply by k-bx** [reply]
@nick-kang I just have two separate objects in my `Env` type, one for sqlx and another for sea-orm. Configured to create zero connections by default, it's sort of ok for now.

**Reply by nick-kang** [reply]
Thanks. Updated to something like this to support using sqlx transactions:

```rust
        let pool = sqlx::Pool::<sqlx::postgres::Postgres>::connect(&database_url)
            .await
            .expect("to connect to database");

        let primary_seaorm: DatabaseConnection =
            SqlxPostgresConnector::from_sqlx_postgres_pool(pool.clone());

        Self {
            primary_seaorm,
            primary_sqlx: pool,
        }
```

**Reply by tyt2y3** [reply]
Or, you can add a method to the State object and construct the SeaORM Connection on-the-fly. It is not expensive.
